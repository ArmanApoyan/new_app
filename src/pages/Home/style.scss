.toDo {
  width: 40%;
  max-width: 400px;
  min-width: 230px;
  height: 20%;
  max-height: 800px;
  scroll-behavior: auto;
  overflow-y: auto;
  overflow-x: hidden;
  &::-webkit-scrollbar {
    width: 6px;
  }
  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 5px grey;
    border-radius: 10px;
  }
  &::-webkit-scrollbar-thumb {
    background: rgb(51, 134, 139);
    border-radius: 20px;
  }
  &::-webkit-scrollbar-thumb:hover {
    background: #3698b1;
  }
  background-color: rgb(255, 255, 255);
  display: flex;
  gap: 20px;
  flex-direction: column;
  align-items: center;
  padding: 20px 0;
  border-radius: 10px;
  box-shadow: 8px 8px 10px rgb(160, 160, 160);

  h2 {
    box-sizing: border-box;
    text-align: left;
    width: 100%;
    padding-left: 20px;
    font-family:ubun;
    font-size: 25px;
  }
}
