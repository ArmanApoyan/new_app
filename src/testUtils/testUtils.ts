import { combineReducers, createStore } from "redux";
import { myReducer } from "../store/Task/reducer";


export function createTestStore() {
    const store = createStore(
      combineReducers({
        task: myReducer,
      })
    );
    return store;
  }