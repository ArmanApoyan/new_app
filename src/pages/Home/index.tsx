import { DragDropContext, Droppable } from "react-beautiful-dnd";
import { Column, Goal, State } from "../../types/global";
import { useDispatch, useSelector } from "react-redux";
import { UPDATE } from "../../store/Task/types";
import Element from "../../components/Element";
import { reorder } from "../../utils";
import { useCallback, useEffect } from "react";
import "./style.scss";
import { useNavigate } from "react-router-dom";
import { action1, getTasks, reorderTasks } from "../../store/Task/action";

const Home: React.FC = () => {
  const navigate = useNavigate()
  useEffect(()=>{
    if(!localStorage.token){
      navigate("/log")
    }
    // @ts-ignore
    dispatch(getTasks())
  },[])
  
  const dispatch = useDispatch();
  const { goals, search, columns } = useSelector((state: State) => state.task);

  function dragEnd(res: any) {
    let data: Goal[] = reorder(
      goals,
      res.source.index,
      res.destination.index,
      res.source.droppableId,
      res.destination.droppableId
    );
     // @ts-ignore
    dispatch(reorderTasks("create", data));
  }

  const handleSearch = useCallback((search: string) => {
    return (elem: Goal) => {
      if(search.length <= 2) {
        return elem
      }
      else if(elem.title.toLowerCase().includes(search.toLowerCase())) {
        return elem
      }
      return false
    }
  }, [])
  
  return (
    <DragDropContext onDragEnd={dragEnd}>
      {columns.map((elm: Column, index: number) => {
        return (
          <Droppable droppableId={elm.title} key={index}>
            {(provided,snapshot) => (
              <div
              data-testid = "column"
                {...provided.droppableProps}
                ref={provided.innerRef}
                className={snapshot.isUsingPlaceholder?"columns dragColumn":"columns"}
              >
                <h2>{elm.title}</h2>
                {goals.filter(handleSearch(search)).map((el: Goal, i: number) => {
                  if (el.status == elm.title) {
                    return <Element key={`${el.id}/${elm.title}`} task={el} index={i} />;
                  }
                })}
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        );
      })}
    </DragDropContext>
  );
};

export default Home;
