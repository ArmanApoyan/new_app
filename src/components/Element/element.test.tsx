import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter, Router } from "react-router-dom";
import Home from "../../pages/Home";
import { createTestStore } from "../../testUtils/testUtils";
import { createMemoryHistory } from "history";

describe("Element component test", () => {
  let store = createTestStore();
  it("Element rendering", () => {
    render(
      <MemoryRouter initialEntries={[{ pathname: "/", search: "task=1" }]}>
        <Provider store={store}>
          <Home />
        </Provider>
      </MemoryRouter>
    );
    expect(screen.getAllByTestId("task")).toMatchSnapshot();
  });
  it("close button click", () => {
    const { container } = render(
      <MemoryRouter initialEntries={[{ pathname: "/", search: "task=1" }]}>
        <Provider store={store}>
          <Home />
        </Provider>
      </MemoryRouter>
    );
    const buttons = screen.getAllByTestId("close");
    fireEvent.click(buttons[0]);
    expect(buttons[0]).not.toBeInTheDocument();
  });
  it("dblclick on task changing url search", () => {
    const history = createMemoryHistory({ initialEntries: ["/"] });
    render(
      <Router location={history.location} navigator={history}>
        <Provider store={store}>
            <Home />
        </Provider>
      </Router>
    );
    const tasks = screen.getAllByTestId("task");
    fireEvent.dblClick(tasks[0]);
    expect(history.location.search).toBe("?task=2");
  });
  it("close button click", () => {
    const { container } = render(
      <MemoryRouter initialEntries={[{ pathname: "/", search: "task=1" }]}>
        <Provider store={store}>
          <Home />
        </Provider>
      </MemoryRouter>
    );
    const button = screen.getAllByTestId("edit_btn");
    fireEvent.click(button[0]);
    expect(screen.getByTestId("modal")).toMatchSnapshot()
  });
  it("dblclick on task opening modal", () => {
    const { container } = render(
      <MemoryRouter initialEntries={[{ pathname: "/", search: "task=1" }]}>
        <Provider store={store}>
          <Home />
        </Provider>
      </MemoryRouter>
    );
    const tasks = screen.getAllByTestId("task");
    fireEvent.dblClick(tasks[0]);
    expect(screen.getByTestId("modal")).toMatchSnapshot()
  });
  it("modal outside click", () => {
    const { container } = render(
      <MemoryRouter initialEntries={[{ pathname: "/", search: "task=1" }]}>
        <Provider store={store}>
          <Home />
        </Provider>
      </MemoryRouter>
    );
    const tasks = screen.getAllByTestId("task");
    fireEvent.dblClick(tasks[0]);
    const body =  screen.getByTestId("body");
    fireEvent.click(body)
    expect(body).toBeInTheDocument()   
  });
});
