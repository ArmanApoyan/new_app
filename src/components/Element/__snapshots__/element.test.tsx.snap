// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element component test Element rendering 1`] = `
Array [
  <div
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    class="goal"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="1"
    data-testid="task"
    draggable="false"
    role="button"
    tabindex="0"
  >
    <h1
      data-testid="title"
    >
      Task 1
    </h1>
    <h3>
      Description 1
    </h3>
    <button
      class="x"
      data-testid="close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <button
      class="edit"
      data-testid="edit_btn"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.854.146a.5.5 0 0 0-.707 0L10.5 1.793 14.207 5.5l1.647-1.646a.5.5 0 0 0 0-.708l-3-3zm.646 6.061L9.793 2.5 3.293 9H3.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.207l6.5-6.5zm-7.468 7.468A.5.5 0 0 1 6 13.5V13h-.5a.5.5 0 0 1-.5-.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.5-.5V10h-.5a.499.499 0 0 1-.175-.032l-.179.178a.5.5 0 0 0-.11.168l-2 5a.5.5 0 0 0 .65.65l5-2a.5.5 0 0 0 .168-.11l.178-.178z"
        />
      </svg>
    </button>
  </div>,
  <div
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    class="goal"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="2"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="2"
    data-testid="task"
    draggable="false"
    role="button"
    tabindex="0"
  >
    <h1
      data-testid="title"
    >
      Task 2
    </h1>
    <h3>
      Description 2
    </h3>
    <button
      class="x"
      data-testid="close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <button
      class="edit"
      data-testid="edit_btn"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.854.146a.5.5 0 0 0-.707 0L10.5 1.793 14.207 5.5l1.647-1.646a.5.5 0 0 0 0-.708l-3-3zm.646 6.061L9.793 2.5 3.293 9H3.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.207l6.5-6.5zm-7.468 7.468A.5.5 0 0 1 6 13.5V13h-.5a.5.5 0 0 1-.5-.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.5-.5V10h-.5a.499.499 0 0 1-.175-.032l-.179.178a.5.5 0 0 0-.11.168l-2 5a.5.5 0 0 0 .65.65l5-2a.5.5 0 0 0 .168-.11l.178-.178z"
        />
      </svg>
    </button>
  </div>,
  <div
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    class="goal"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="3"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="3"
    data-testid="task"
    draggable="false"
    role="button"
    tabindex="0"
  >
    <h1
      data-testid="title"
    >
      Task 3
    </h1>
    <h3>
      Description 3
    </h3>
    <button
      class="x"
      data-testid="close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <button
      class="edit"
      data-testid="edit_btn"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.854.146a.5.5 0 0 0-.707 0L10.5 1.793 14.207 5.5l1.647-1.646a.5.5 0 0 0 0-.708l-3-3zm.646 6.061L9.793 2.5 3.293 9H3.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.207l6.5-6.5zm-7.468 7.468A.5.5 0 0 1 6 13.5V13h-.5a.5.5 0 0 1-.5-.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.5-.5V10h-.5a.499.499 0 0 1-.175-.032l-.179.178a.5.5 0 0 0-.11.168l-2 5a.5.5 0 0 0 .65.65l5-2a.5.5 0 0 0 .168-.11l.178-.178z"
        />
      </svg>
    </button>
  </div>,
]
`;

exports[`Element component test close button click 1`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="body"
    data-testid="body"
  >
    <button
      class="close"
      data-testid="btn"
    >
      X
    </button>
    <form
      class="newTask"
    >
      <select
        name="status"
      >
        <option
          value="To Do"
        >
          To Do
        </option>
        <option
          value="On Progress"
        >
          On Progress
        </option>
        <option
          value="Done"
        >
          Done
        </option>
      </select>
      <input
        data-testid="inp"
        name="title"
        placeholder="Title"
        type="text"
        value="Task 2"
      />
      <textarea
        data-testid="textarea"
        name="description"
        placeholder="Description"
      >
        Description 2
      </textarea>
      <button
        class="submit"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;

exports[`Element component test dblclick on task opening modal 1`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="body"
    data-testid="body"
  >
    <button
      class="close"
      data-testid="btn"
    >
      X
    </button>
    <div
      class="view"
    >
      <p
        class="id"
        data-testid="p"
      >
        2
      </p>
      <p
        class="status"
        data-testid="p"
      >
        On Progress
      </p>
      <p
        class="title"
        data-testid="p"
      >
        Task 2
      </p>
      <p
        class="des"
        data-testid="p"
      >
        Description 2
      </p>
    </div>
  </div>
</div>
`;

exports[`Element component test dblclick on task opening modal 2`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="body"
    data-testid="body"
  >
    <button
      class="close"
      data-testid="btn"
    >
      X
    </button>
    <div
      class="view"
    >
      <p
        class="id"
        data-testid="p"
      >
        2
      </p>
      <p
        class="status"
        data-testid="p"
      >
        On Progress
      </p>
      <p
        class="title"
        data-testid="p"
      >
        Task 2
      </p>
      <p
        class="des"
        data-testid="p"
      >
        Description 2
      </p>
    </div>
  </div>
</div>
`;

exports[`Element component test modal 1`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="body"
    data-testid="body"
  >
    <button
      class="close"
      data-testid="btn"
    >
      X
    </button>
    <div
      class="view"
    >
      <p
        class="id"
        data-testid="p"
      >
        2
      </p>
      <p
        class="status"
        data-testid="p"
      >
        On Progress
      </p>
      <p
        class="title"
        data-testid="p"
      >
        Task 2
      </p>
      <p
        class="des"
        data-testid="p"
      >
        Description 2
      </p>
    </div>
  </div>
</div>
`;
